Fix bugs with Create Index Pattern wizard loading state and createReasonableWait UX helper

cjcenizal opened the pull request:
Bugs fixed:

* Step 1 and EmptyState no longer appear simultaneously
![image](https://user-images.githubusercontent.com/1238659/36613810-7a5d6f5e-188f-11e8-8968-9fec00edfe65.png)
* The Refresh button in the EmptyState works now
* Icons updated
* Fixed React error with `isDisabled` being used on `option` element
* Tweaked `createReasonableWait` to ensure certain requests take a minimum amount of time to avoid UI flicker

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/200/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/205/)

chrisronline commented:
Do you think it makes sense to tack on #16959 to this? @cjcenizal

cjcenizal commented:
@chrisronline I think I'd rather do that in a separate PR.

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/lib/create_reasonable_wait.js:
```diff
@@ -1,3 +1,12 @@
-export function createReasonableWait(cb) {
-  return setTimeout(cb, 100);
+export function createReasonableWait(...promises) {
```
What if we renamed this to `ensureMinimumWait`?

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/lib/create_reasonable_wait.js:
```diff
@@ -1,3 +1,12 @@
+      // is happening.
+      setTimeout(() => {
+        resolve();
+      }, 500);
```
It might be nice if this were configurable. Maybe change the parameters to accept:

1) An array of promises
2) A timeout (default to 300)

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/lib/__tests__/create_reasonable_wait.test.js:
```diff
@@ -1,10 +1,11 @@
-    const callback = sinon.spy();
-    createReasonableWait(callback);
-    expect(callback.notCalled).toBeTruthy();
+  it('resolves all promises passed to it', async () => {
```
What if we did this for tests?

```
import { createReasonableWait } from '../create_reasonable_wait';

jest.useFakeTimers();

const nextTick = () => new Promise(res => process.nextTick(res));

describe('createReasonableWait', () => {
  it('resolves all promises passed to it', (done) => {
    const promiseA = new Promise(resolve => resolve('a'));
    const promiseB = new Promise(resolve => resolve('b'));
    createReasonableWait(promiseA, promiseB)
      .then(([a, b]) => {
        expect(a).toBe('a');
        expect(b).toBe('b');
        done();
      });
    jest.runAllTimers();
  });

  describe('timing', () => {
    it('should not finish if the min time has not been reached', async () => {
      const results = [];

      const promiseA = new Promise(resolve => setTimeout(() => {
        results.push('a');
        resolve();
      }, 100));

      const promiseB = new Promise(resolve => setTimeout(() => {
        results.push('b');
        resolve();
      }, 100));

      createReasonableWait(promiseA, promiseB).then(() => {
        results.push('complete');
      })

      jest.advanceTimersByTime(499);
      await nextTick();

      expect(results).toEqual(['a', 'b']);
    });

    it('should finish if the min time has been reached', async () => {
      const results = [];

      const promiseA = new Promise(resolve => setTimeout(() => {
        results.push('a');
        resolve();
      }, 100));

      const promiseB = new Promise(resolve => setTimeout(() => {
        results.push('b');
        resolve();
      }, 100));

      createReasonableWait(promiseA, promiseB).then(() => {
        results.push('complete');
      })

      jest.advanceTimersByTime(500);
      await nextTick();

      expect(results).toEqual(['a', 'b', 'complete']);
    });
  });


});
```

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/components/step_index_pattern/__tests__/__snapshots__/step_index_pattern.test.js.snap:
```diff
@@ -11,52 +11,17 @@ exports[`StepIndexPattern should disable the next step if the index pattern exis
     goToNextStep={[Function]}
     isInputInvalid={false}
-    isNextStepDisabled={false}
+    isNextStepDisabled={true}
```
This doesn't look right.

Because of the addition of extra Promise layer, we need to perform another `await` in the test.

The test should look like:
```
it('should disable the next step if the index pattern exists', async () => {
    const component = shallow(
      <StepIndexPattern
        allIndices={allIndices}
        isIncludingSystemIndices={false}
        esService={esService}
        savedObjectsClient={{
          find: () => ({ savedObjects: [
            { attributes: { title: 'k*' } }
          ] })
        }}
        goToNextStep={goToNextStep}
        initialQuery="k*"
      />
    );

    // Allow the componentWillMount code to execute
    // https://github.com/airbnb/enzyme/issues/450
    await component.update(); // Fire `componentWillMount()`
    await component.update(); // Force update the component post async actions
    await component.update(); // Ensure the reasonable wait finishes

    expect(component).toMatchSnapshot();
  });
```

chrisronline replied:
Updated this comment @cjcenizal

chrisronline replied:
For #2, the code is https://github.com/elastic/kibana/pull/16895/files#diff-ac3256ee9799125617e80b1b906d9d78R51

chrisronline replied:
Also, see #16968

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/components/step_index_pattern/__tests__/__snapshots__/step_index_pattern.test.js.snap:
```diff
@@ -96,52 +61,17 @@ exports[`StepIndexPattern should properly fetch indices for the initial query 1`
     goToNextStep={[Function]}
     isInputInvalid={false}
-    isNextStepDisabled={false}
+    isNextStepDisabled={true}
```
Ditto

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/components/step_index_pattern/__tests__/__snapshots__/step_index_pattern.test.js.snap:
```diff
@@ -218,52 +148,17 @@ exports[`StepIndexPattern should render some indices 1`] = `
-    }
-    query="k*"
-  />
+  <LoadingIndices />
```
Ditto

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/__tests__/create_index_pattern_wizard.test.js:
```diff
@@ -73,10 +67,16 @@ describe('CreateIndexPatternWizard', () => {
 
+    component.setState({
+      isInitiallyLoadingIndices: false,
+      allIndices: [{}],
```
We should probably add an index here as the test says `when there are indices`

cjcenizal replied:
I was thinking the empty object acts as an index. Were you thinking we should add some props to it?

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/lib/create_reasonable_wait.js:
```diff
@@ -1,3 +1,12 @@
-export function createReasonableWait(cb) {
-  return setTimeout(cb, 100);
+export function createReasonableWait(...promises) {
+  return Promise.all([
```
Per our conversation, I'm thinking this might be slightly more readable:
```
async function ensureMinimumTime(block, minimumTimeMs = 300) {
  // console.log(`ensureMinimumTime() minimumTimeMs=${minimumTimeMs}`);
  const start = +new Date();
  await block();
  const end = +new Date();
  const duration = end - start;

  if (duration < minimumTimeMs) {
    const timeToAdd = minimumTimeMs - (end - start);
    await new Promise(resolve => setTimeout(resolve, timeToAdd));
  }
}
```

Feel free to clean up the variable naming, etc

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/373/)

cjcenizal commented:
@chrisronline I implemented the new `ensureMinimumTime` helper and drastically simplified the tests (and incorporated your suggestions). Can you take another look?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/384/)

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/components/step_index_pattern/__tests__/step_index_pattern.test.js:
```diff
@@ -23,166 +20,75 @@ const savedObjectsClient = {
-    await instance.onQueryChanged({
-      target: { value: 'k' }
+describe('StepIndexPattern', () => {
+  describe('step 1', () => {
```
Does this extra describe feel necessary? When I see this, I assume there is a step 2 or something

cjcenizal replied:
You're right, this is totally superfluous.

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/components/step_index_pattern/__tests__/step_index_pattern.test.js:
```diff
@@ -23,166 +20,75 @@ const savedObjectsClient = {
+    it('renders the loading state', () => {
+      const component = createComponent();
+      component.setState({ isLoadingIndices: true });
+      expect(component.find('[data-test-subj="createIndexPatternStep1Loading"]')).toMatchSnapshot();
```
Hmm. So it looks like you changed these to not look at the entire component, but rather a subset of the component. I'm wondering if that's what we want since we don't have a single test ensuring the overall structure is maintained (like the component wrapped in a `<EuiPanel>` or the ordering makes sense). Can you elaborate more on what you're thinking with this change?

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/lib/__tests__/ensure_minimum_time.test.js:
```diff
@@ -0,0 +1,28 @@
+    const promise = new Promise(resolve => resolve());
+    await ensureMinimumTime(promise, 100);
+    const endTime = new Date().getTime();
+    expect(endTime - startTime).toBeGreaterThanOrEqual(100);
```
I don't really know why yet, but I see this sometimes with this test:
```
● ensureMinimumTime › resolves in the amount of time provided, at minimum

    expect(received).toBeGreaterThanOrEqual(expected)
Expected value to be greater than or equal:
      100
    Received:
      99
```

chrisronline replied:
It's possible it only happens in watch mode when running the jest tests but it's still weird...

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/428/)
